tests/tt_metal/tt_metal/test_bfp4_conversion.cpp

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <algorithm>
+#include <functional>
+#include <random>
+
+#include "tt_metal/host_api.hpp"
+#include "common/bfloat4.hpp"
+#include "common/bfloat16.hpp"
+#include "common/test_tiles.hpp"
+#include "stdio.h"
+
+using namespace tt;
+
+int main(int argc, char **argv) {
+    bool pass = true;
+
+    try {
+
+        uint32_t single_bfp4_tile_size = tile_size(tt::DataFormat::Bfp4_b);
+        uint32_t num_tiles = 1;
+        uint32_t size_in_bytes = num_tiles * single_bfp4_tile_size;
+
+        int num_float_in_tile = 1024;
+        int float_data_size = num_tiles * num_float_in_tile;
+
+        std::vector<float> fp32_vec(float_data_size, 0);
+        for (int i = 0; i < fp32_vec.size(); i++) {
+            fp32_vec.at(i) = i;
+        }
+
+        std::vector<uint32_t> shape_vec = {1, num_tiles, 32, 32};
+        std::vector<float> tiled_fp32_vec = convert_layout(fp32_vec, shape_vec, TensorLayout::LIN_ROW_MAJOR, TensorLayout::TILED32_4FACES);
+
+        std::vector<uint32_t> packed_bfp4b_tile_vec_rm_in = pack_fp32_vec_as_bfp4_tiles(fp32_vec, /*row_major_input=*/true, /*is_exp_a=*/false);
+        std::vector<float> unpacked_bfp4b_tile_vec_rm_out = unpack_bfp4_tiles_into_float_vec(packed_bfp4b_tile_vec_rm_in, /*row_major_output*/true, /*is_exp_a=*/false);
+
+        std::vector<uint32_t> packed_bfp4b_tile_vec_tile_in = pack_fp32_vec_as_bfp4_tiles(tiled_fp32_vec, /*row_major_input=*/false, /*is_exp_a=*/false);
+        std::vector<float> unpacked_bfp4b_tile_vec_tile_out = unpack_bfp4_tiles_into_float_vec(packed_bfp4b_tile_vec_tile_in, /*row_major_output=*/false, /*is_exp_a=*/false);
+
+
+        // ////////////////////////////////////////////////////////////////////////////
+        // //                      Validation
+        // ////////////////////////////////////////////////////////////////////////////
+        std::vector<float> tiled_to_rm_fp32_vec = convert_layout(unpacked_bfp4b_tile_vec_tile_out, shape_vec, TensorLayout::TILED32_4FACES, TensorLayout::LIN_ROW_MAJOR);
+        std::vector<float> rm_to_tiled_fp32_vec = convert_layout(unpacked_bfp4b_tile_vec_rm_out, shape_vec, TensorLayout::LIN_ROW_MAJOR, TensorLayout::TILED32_4FACES);
+
+        // Ensure that passing in row_major_input=true and row_major_output=true are inverses of row_major_input=false and row_major_output=false yield the same result
+        pass &= (packed_bfp4b_tile_vec_rm_in == packed_bfp4b_tile_vec_tile_in);
+
+        printf("MT: unpacked_bfp4b_tile_vec_rm_out.size() = %ld\n", unpacked_bfp4b_tile_vec_rm_out.size());
```

Comment:
there are a few of these `printf` statements, do you mind just deleting all of these before merging?